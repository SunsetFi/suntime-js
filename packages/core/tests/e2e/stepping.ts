import { describe, it, expect } from "vitest";

import { StaticJsRealm } from "../../src/index.js";

describe("E2E: Stepping", () => {
  it("Steps through opts individually", async () => {
    let iterations = 0;

    const realm = StaticJsRealm({
      runTask(task) {
        while (true) {
          iterations++;
          const { done } = task.next();
          if (done) {
            break;
          }
        }
      },
    });

    const code = `
      for(let i = 0; i < 10; i++) {};
    `;

    await realm.evaluateScript(code);

    // Ensure that we are stepping through individual operations and not consuming
    // everything on one next() call.
    expect(iterations).toBeGreaterThan(10);
    expect(
      `Took ${iterations} iterations to run the script.`,
    ).toMatchSnapshot();
  });

  it("Records the line and column of each step", async () => {
    const steps: string[] = [];
    const realm = StaticJsRealm({
      runTask(task) {
        while (true) {
          // The function doesn't start until the first next() call, so we are at -1/-1 right now.
          const { done } = task.next();
          if (done) {
            break;
          }

          // Now that we ran, we should be paused at an operation.
          const op = task.operation!;
          steps.push(
            `Line: ${op.location.start.line}, Column: ${op.location.start.column}`,
          );
        }
      },
    });

    const code = `
      for(let i = 0; i < 10; i++) {};
    `;

    await realm.evaluateScript(code);

    expect(steps).toMatchSnapshot();
  });
});
