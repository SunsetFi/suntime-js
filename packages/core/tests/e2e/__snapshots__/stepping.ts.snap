// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`E2E: Stepping > Records the line and column of each step 1`] = `
[
  "Line: 1, Column: 0",
  "Line: 2, Column: 6",
  "Line: 2, Column: 10",
  "Line: 2, Column: 18",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 34",
  "Line: 2, Column: 29",
  "Line: 2, Column: 21",
  "Line: 2, Column: 21",
  "Line: 2, Column: 25",
  "Line: 2, Column: 36",
]
`;

exports[`E2E: Stepping > Steps through opts individually 1`] = `"Took 62 iterations to run the script."`;
